src/Bounds.cpp

```diff
@@ -1237,12 +1237,19 @@ class Bounds : public IRVisitor {
 
         if (op->is_intrinsic(Call::abs)) {
             Interval a = arg_bounds.get(0);
-            interval.min = make_zero(t);
+
+            if (a.has_lower_bound()) {
+                interval.min = cast(t, Max::make(make_zero(a.min.type()), a.min));
+            } else {
+                interval.min = make_zero(t);
+            }
+
             if (a.is_bounded()) {
                 if (equal(a.min, a.max)) {
                     interval = Interval::single_point(Call::make(t, Call::abs, {a.max}, Call::PureIntrinsic));
                 } else if (op->args[0].type().is_int() && op->args[0].type().bits() >= 32) {
-                    interval.max = Max::make(Cast::make(t, -a.min), Cast::make(t, a.max));
+                    interval.min = Cast::make(t, Max::make(make_zero(a.min.type()), Max::make(a.min, -a.max)));
+                    interval.max = Cast::make(t, Max::make(-a.min, a.max));
```

Comment:
It's abs, so t is unsigned, and a.min.type() is signed. Don't you want to do the max in the unsigned type?

Reply:
If a = [2, 10] then originally you got `interval.max = max(u32(-2), u32(10)) = 4294967294`

Reply:
Right, I was only considering the a.min < 0 case. Doing it in the unsigned type would be: `max(abs(a.min), abs(a.max))`, but I'm guessing elsewhere in bounds inference isn't going to like that.

Reply:
Yeah, the original code works for `a.min < 0`, but I think the new code works for both negative and positive `a.min`. If you assume negation can't overflow, then the comparison should be done in the signed type.